### ISSUE TITLE
[Question for help] Unit test with EF Core and ExtCore

### ISSUE THREAD

**Issue by postb99** [ExtCore/ExtCore#91]
Hi Dmitry,

I told myself that it would be a good learning curve to write an XUnit test to call my code that performs inserts with EF Core. However, it has access to ExtCore EF IStorage object.

I tried to make something functional but my unit test program fails at finding the repository implementation.

I admit that for later tests I will not need real database access (can go with mocks or just write the code another way).

By the way, do you have unit tests for ExtCore?

I hope that my help request is not too off-topic.

Reminder of my repo's branch: https://github.com/OSAMES/Base/tree/feature/security_extension

PS: I'll have a look at David Norena's test program. 

My stack trace: 

System.NullReferenceException : Object reference not set to an instance of an object.
   at ExtCore.Infrastructure.ExtensionManager.GetImplementations[T](Func`2 predicate, Boolean useCaching)
   at ExtCore.Infrastructure.ExtensionManager.GetInstances[T](Func`2 predicate, Boolean useCaching, Object[] args)
   at ExtCore.Infrastructure.ExtensionManager.GetInstance[T](Func`2 predicate, Boolean useCaching, Object[] args)
   at ExtCore.Data.EntityFramework.Storage.GetRepository[TRepository]()

**Comment by DmitrySikorsky** [comment]
I don't use unit tests :( There is nothing to be proud, but...

What version of the ExtCore do you use? I have fixed that issue in one of the last releases, now you  should just have null repository, not null reference exception. That was because of this: https://github.com/ExtCore/ExtCore/blob/master/src/ExtCore.Data.EntityFramework/Storage.cs#L36 (I added null check).

Could you show me the place where you have that exception? I will try to help.

**Comment by postb99** [comment]
Hi Dmitry,

I use ExtCore 3.0, always following your latest available nuGet package. Please then have a look when you have time to my test project (https://github.com/OSAMES/Base/tree/feature/security_extension/Tests/SecurityTest) to validate the approach, it could be my misunderstanding on how to set up things together. 

I have read some things about mocking the EF so that it uses in-memory database, I may consider it later, but for this very specific test I was looking for a quick way to launch again my database initialization code. Anyways, this morning I achieve fixing the database so the test will become useless :-p. Yesterday I was also looking to log the SQL queries generated by EF but didn't find it how (the right logger to set log level of), a bit frustratring but this is off-topic :-)

**Comment by postb99** [comment]
Well, I did some progresses but still feel stuck about your fix of #85 I obviously don't have in the 3.0.0 release because the repository is null after lookup by ExtensionManager. But why it is null? I fed ExtensionManager whith the correct assemblies I think, this was this morning's progress.

Thanks a lot for helping, for now please run the test in DatabaseInitializerTest (base code at https://github.com/OSAMES/Base/tree/feature/security_extension/Tests/SecurityTest), it will faill because of the repository not being found.

**Comment by postb99** [comment]
Hi Dmitry,

I'm glad to inform you that I succeeded in setting up things, please can you review my work, maybe some classes could move into an ExtCore.Tests library? The IRequestHandler interface I picked up from Platformus seems interesting to me for any ExtCore based web application, doesn't it? I also have additional small questions.

I created a [TestContext](https://github.com/OSAMES/Base/blob/feature/security_extension/Tests/SecurityTest/TestContext.cs) class that implements the interface that I use by DI in the application in the controllers/modules.

In this class I define which provider to use, here sqlite, as I would do in web application's startup.

And that was just all! I like when learning ends up to minimal code writing.

My questions are the following:
- maybe is there some class I could use instead of creating my own implementation for IOptions<StorageContextOptions>? I had a look into ExtCore.
- I have trouble with the Sqlite db path, so I ended up as building an absolute path. It seems that it's the way to go. At least it works (DatabaseInitializerTest test passes, the other test class in a WIP).

**Comment by DmitrySikorsky** [comment]
Hi @postb99! Sorry for the delay. I'm not very good with tests, so I just don't know how to help without learning unit testing :) I'm not sure I understand the first question about the options. Could you describe it more? And what problem do you have with the SQLite db path?

**Comment by postb99** [comment]
Hi Dmitry, I can help you learning, I do a lot of C# unit testing.
To run the unit test, I personally use Resharper and click on the icon it provides in the unit test code, but you can use [XUnit console runner](https://xunit.github.io/docs/getting-started-desktop.html#add-xunit-runner-ref).

When tests starts, code of DatabaseFixture's constructor is called because my test class is decorated with "[Collection("Database collection")]" (I learned this, [Xunit docs](https://xunit.github.io/docs/shared-context.html) are well written).

My DatabaseFixture exposes a DatabaseContext property that provides access to IStorage.

- My first question related to [this code](https://github.com/OSAMES/Base/blob/feature/security_extension/Tests/SecurityTest/TestContext.cs#L29) is just a notice that I implemented the IOptions<StorageContextOptions> as [TestOptions](https://github.com/OSAMES/Base/blob/feature/security_extension/Tests/SecurityTest/TestContext.cs#L45) class, but asked myself whether I could find an already existing implementation.

- My second question related to [this code](https://github.com/OSAMES/Base/blob/feature/security_extension/Tests/SecurityTest/TestContext.cs#L26) where I format an absolute path, because the relative path failed, I was not sure of the root and didn't give many tries. On SO I saw people using absolute path but I don't like this much, so was wondering about this.

The interest of unit testing is that you have to worry about the interfaces that are used by DI and implement them a way or another in the unit tests layer. It took me some time understanding this fully, also because I have a child interface (IStorage).

May I ask @DavidNorena whether he's also smart in unit testing? My two technical questions were in fact side questions, going deeper into good-looking code, but it's the first time I work with a webapp with DI so I'm enthusiastic about what I learned because I'm also interested of getting feedback if I do something the wrong way. Sorry for the long writing :-)

Now I'm looking to adding EF Core SQL queries logging so that I can fix my code, I'm new to EF. I'll see that tomorrow...

**Comment by DavidNorena** [comment]
Hi @postb99, sorry but I'm starting to learn Unit Testing too, so sad, I cant help you with that .... :(

**Comment by postb99** [comment]
Hi @DavidNorena , thanks for the reply. In fact, I have a working unit test so feel free to learn. I've merged it to my master branch ([test classes](https://github.com/OSAMES/Base/tree/master/Tests/SecurityTest)). XUnit is the test framework to go with, supported by MS for .NET Core.

I also found out how to log EF Core verbosely with the SQL queries. The configuration method is [here](https://github.com/OSAMES/Base/blob/master/Tests/SecurityTest/TestContext.cs#L66) with just an additional class to create. Taken from MS docs :-)

Happy learning. I close the issue because things are satisfying for most parts.
